Content.Server/Revenant/EntitySystems/RevenantSystem.cs

```diff
@@ -0,0 +1,198 @@
+    private void OnStatusAdded(EntityUid uid, RevenantComponent component, StatusEffectAddedEvent args)
+    {
+        if (!TryComp<AppearanceComponent>(uid, out var app))
+            return;
+
+        if (args.Key == "Stun")
+            app.SetData(RevenantVisuals.Stunned, true);
+    }
+
+    private void OnStatusEnded(EntityUid uid, RevenantComponent component, StatusEffectEndedEvent args)
+    {
+        if (!TryComp<AppearanceComponent>(uid, out var app))
+            return;
+
+        if (args.Key == "Stun")
+            app.SetData(RevenantVisuals.Stunned, false);
+        else if (args.Key == "Corporeal")
+            _movement.RefreshMovementSpeedModifiers(uid);
+    }
+
+    private void OnExamine(EntityUid uid, RevenantComponent component, ExaminedEvent args)
+    {
+        if (args.Examiner == args.Examined)
+        {
+            args.PushMarkup(Loc.GetString("revenant-essence-amount",
+                ("current", Math.Round(component.Essence)), ("max", Math.Round(component.EssenceRegenCap))));
+        }
+    }
+
+    private void OnDamage(EntityUid uid, RevenantComponent component, DamageChangedEvent args)
+    {
+        if (!HasComp<CorporealComponent>(uid) || args.DamageDelta == null)
+            return;
+
+        var essenceDamage = args.DamageDelta.Total.Float() * component.DamageToEssenceCoefficient * -1;
+        ChangeEssenceAmount(uid, essenceDamage, component);
+    }
+
+    public bool ChangeEssenceAmount(EntityUid uid, float amount, RevenantComponent? component = null, bool allowDeath = true, bool regenCap = false)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (!allowDeath && component.Essence + amount <= 0)
+            return false;
+
+        component.Essence += amount;
+
+        if (regenCap)
+            component.Essence = Math.Min(component.Essence, component.EssenceRegenCap);
+
+        _alerts.ShowAlert(uid, AlertType.Essence, (short) Math.Clamp(Math.Round(component.Essence / 10f), 0, 16));
+
+        if (component.Essence <= 0)
+        {
+            component.Essence = component.EssenceRegenCap;
+            _polymorphable.PolymorphEntity(uid, "Ectoplasm");
+        }
+
+        UpdateUserInterface(component);
+        return true;
+    }
+
+    private bool CanUseAbility(EntityUid uid, RevenantComponent component, float abilityCost, Vector2 debuffs)
+    {
+        if (!ChangeEssenceAmount(uid, abilityCost, component, false))
+        {
+            _popup.PopupEntity(Loc.GetString("revenant-not-enough-essence"), uid, Filter.Entities(uid));
+            return false;
+        }
+
+        var tileref = Transform(uid).Coordinates.GetTileRef();
+        if (tileref != null)
+        {
+            if(_physics.GetEntitiesIntersectingBody(uid, (int) CollisionGroup.Impassable).Count > 0)
+            {
+                _popup.PopupEntity(Loc.GetString("revenant-in-solid"), uid, Filter.Entities(uid));
+                return false;
+            }
+        }
```

Comment:
Does this change the amount and then check if you're in a solid? Seems like you might be better checking if they have enough, then checking solid, then deducting?